Fix issue #618: Omnipod errors should display localized Strings

itsmojo opened the pull request:
* Return the underlying localized error string for PodCommError.commsError
* Have all OmnipodPumpManager commands return a localized error
* Fix coding errors in CommandResponseViewController for Read Pulse Log
and Read Pulse Log which return a String and not an error

itsmojo committed:
Fix issue #618: Omnipod commands errors should display localized strings when possible
* Return the underlying localized error string for PodCommError.commsError
* Have all OmnipodPumpManager commands return a localized error
* Fix coding errors in CommandResponseViewController for Read Pulse Log
and Read Pulse Log which return a String and not an error

ps2 commented on OmniKit/PumpManager/OmnipodPumpManager.swift:
```diff
@@ -947,9 +947,10 @@ extension OmnipodPumpManager {
-    public func readPodStatus(completion: @escaping (String?) -> Void) {
-        guard self.hasActivePod else {
-            completion(String(describing: OmnipodPumpManagerError.noPodPaired))
+    public func readPodStatus(completion: @escaping (String) -> Void) {
```
It would be nice to change this to return Result<String, Error>, but we can do that later.

itsmojo replied:
My initial impression was that having a <String, Error> return for OmnipodPumpManager's ReadPodStatus and ReadPulseLog commands might make things more complicated than just returning a String to display (with either the results or the error message), but I would certainly be willing to rework these later if needed now that I better understand how this might look.

ps2 commented on OmniKit/PumpManager/OmnipodPumpManager.swift:
```diff
@@ -1032,15 +1033,26 @@ extension OmnipodPumpManager {
     }
 
-    public func readPulseLog(completion: @escaping (String?) -> Void) {
+    public func readPulseLog(completion: @escaping (String) -> Void) {
```
Same here: Result<String, Error> would be more idiomatic.

itsmojo replied:
Ditto

ps2 commented on OmniKitUI/ViewControllers/CommandResponseViewController.swift:
```diff
@@ -82,7 +84,9 @@ extension CommandResponseViewController {
             pumpManager.playTestBeeps() { (error) in
                 let response: String
-                if let error = error {
+                if let localizedError = error as? LocalizedError, !localizedError.localizedDescription.isEmpty {
```
Just calling error.localizedDescription will work for this. Those PodCommErrors that return nil in errorDescription will return an odd string for `localizedDescription`, so the "isEmpty" will always be false. 

We can add descriptions for those PodCommErrors that have nil errorDescriptions later.

itsmojo replied:
Thanks for the info. Updated code assuming that localizedError.localizedDescription.isEmpty will always be false

ps2 reviewed (changes requested):
Looks good in general. I think we should avoid the checking and downcasting of errors in general, and use localizedDescription for these types of errors that we're simply printing in the testing commands.

itsmojo committed:
Fix issue #617: OmniKitUI displays a mystery single dot for certain setup errors (#621)

* Return the underlying localized error string for PodCommError.commsError
* Improved lastError handling logic in PairPodSetupViewController,
InsertCannulaSetupViewController & ReplacePodViewController to never display a single
dot (".") for any error and to always display some message for any errors returned.

itsmojo committed:
Simplifications because localizedDescription.isEmpty is always false

itsmojo committed:
Improved deactivate pod handling for a faulted pod (#623)

* Improved deactivate pod handling for a faulted pod
* Reconcile insulin values for any interrupted boluses or temp basals
* Do automatic read pulse log as pod messages now persist across pods

* Remove incorrect duplicated code

ps2 commented on OmniKitUI/ViewControllers/CommandResponseViewController.swift:
```diff
@@ -27,15 +27,11 @@ extension CommandResponseViewController {
                         response = messageWithRecovery
                     }
-                } else if let localizedError = error as? LocalizedError, !localizedError.localizedDescription.isEmpty {
+                } else if let localizedError = error as? LocalizedError {
```
I think the whole conditional can be simplified:
```
} else if let localizedError = error as? LocalizedError {
    response = localizedError.localizedDescription
} else if let error = error {
    response = String(describing: error)
}
```
Can be turned into:
```
} else if let error = error {
    response = error.localizedDescription
}
```

itsmojo replied:
Makes sense if localizedDescription.isEmpty is always false

ps2 replied:
I should've mentioned there are several of these in this PR that could be simplified.

itsmojo replied:
Done. Now that PodCommsError.commsError.localizedDescription has a more meaningful content, the simplified version will display a meaningful error message for RileyLink errors with this more simplified method.

itsmojo committed:
Another simplification because localizedDescription.isEmpty will always be false

ps2 committed:
Merge branch 'omnipod-command-error-fixes' of https://github.com/itsmojo/rileylink_ios into itsmojo-omnipod-command-error-fixes

itsmojo committed:
Update CommandResponseViewController.swift

More error simplifications

ps2 committed:
More cleanup

ps2 committed:
Another error string cleanup
